src/core/lib/event_engine/ares_resolver.cc

```diff
@@ -0,0 +1,714 @@
+        addr.sin6_port = htons(hostname_qa->port);
+        result.emplace_back(reinterpret_cast<const sockaddr*>(&addr), addr_len);
+        char output[INET6_ADDRSTRLEN];
+        ares_inet_ntop(AF_INET6, &addr.sin6_addr, output, INET6_ADDRSTRLEN);
+        GRPC_ARES_RESOLVER_TRACE_LOG(
+            "request:%p c-ares resolver gets a AF_INET6 result: \n"
+            "  addr: %s\n  port: %d\n  sin6_scope_id: %d\n",
+            ares_resolver, output, hostname_qa->port, addr.sin6_scope_id);
+        break;
+      }
+      case AF_INET: {
+        size_t addr_len = sizeof(struct sockaddr_in);
+        struct sockaddr_in addr;
+        memset(&addr, 0, addr_len);
+        memcpy(&addr.sin_addr, hostent->h_addr_list[i], sizeof(struct in_addr));
+        addr.sin_family = static_cast<unsigned char>(hostent->h_addrtype);
+        addr.sin_port = htons(hostname_qa->port);
+        result.emplace_back(reinterpret_cast<const sockaddr*>(&addr), addr_len);
+        char output[INET_ADDRSTRLEN];
+        ares_inet_ntop(AF_INET, &addr.sin_addr, output, INET_ADDRSTRLEN);
+        GRPC_ARES_RESOLVER_TRACE_LOG(
+            "request:%p c-ares resolver gets a AF_INET result: \n"
+            "  addr: %s\n  port: %d\n",
+            ares_resolver, output, hostname_qa->port);
+        break;
+      }
+    }
+  }
+  ares_resolver->event_engine_->Run(
+      [callback = std::move(callback), result = std::move(result)]() mutable {
+        callback(std::move(result));
+      });
+}
+
+void AresResolver::OnSRVQueryDoneLocked(void* arg, int status, int /*timeouts*/,
+                                        unsigned char* abuf, int alen) {
+  std::unique_ptr<QueryArg> qa(static_cast<QueryArg*>(arg));
+  auto* ares_resolver = qa->ares_resolver;
+  auto nh = ares_resolver->callback_map_.extract(qa->callback_map_id);
+  GPR_ASSERT(!nh.empty());
+  GPR_ASSERT(
+      absl::holds_alternative<EventEngine::DNSResolver::LookupSRVCallback>(
+          nh.mapped()));
+  auto callback = absl::get<EventEngine::DNSResolver::LookupSRVCallback>(
+      std::move(nh.mapped()));
+  if (status != ARES_SUCCESS) {
+    std::string error_msg = absl::StrFormat(
+        "c-ares status is not ARES_SUCCESS qtype=SRV name=%s: %s",
+        qa->qname.c_str(), ares_strerror(status));
+    GRPC_ARES_RESOLVER_TRACE_LOG("request:%p on_srv_done_locked: %s",
+                                 ares_resolver, error_msg.c_str());
+    ares_resolver->event_engine_->Run(
+        [callback = std::move(callback),
+         status = AresStatusToAbslStatus(status, error_msg)]() mutable {
+          callback(status);
+        });
+    return;
+  }
+  struct ares_srv_reply* reply = nullptr;
+  const int parse_status = ares_parse_srv_reply(abuf, alen, &reply);
+  GRPC_ARES_RESOLVER_TRACE_LOG("request:%p ares_parse_srv_reply: %d",
+                               ares_resolver, parse_status);
+  std::vector<EventEngine::DNSResolver::SRVRecord> result;
+  if (parse_status == ARES_SUCCESS) {
```

Comment:
If it's not ARES_SUCCESS, shouldn't we generate an error status?

Reply:
Great catch! The old code does not: https://source.corp.google.com/piper///depot/google3/third_party/grpc/src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc;l=732-747;bpv=0;bpt=1

But it looks like a bug to me, since in this case we might stay idle and not make any progress until the request timeout or cancel. I changed it so that we generate an error status and report back.

Reply:
Great point! Totally agree with using early return which makes better code structure.